STM32/FLASH_readwrite/Src/FLASH_readwrite.c

```diff
@@ -1,140 +1,248 @@
-// calibration data to the flash. To access these variables the syntax below is used.
-extern uint32_t _FlashSector; // Variable defined in ld linker script.
-extern uint32_t _FlashAddr;   // Variable defined in ld linker script.
-#define FLASH_SECTOR ((uint32_t) &_FlashSector)
-#define FLASH_ADDR   ((uint32_t) &_FlashAddr)
-
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-// Write data to FLASH_ADDR+indx directly. This method does not
-// ensure any data integrity as data validation is performed.
-void writeToFlash(uint32_t indx, uint32_t size, uint8_t *data)
+/***************************************************************************************************
+** PRIVATE FUNCTION PROTOTYPES
+***************************************************************************************************/
+
+static uint32_t getFlashSector(uint32_t address);
+static int isWriteWithinSector(uint32_t address, uint32_t size);
+static int isProgramMemory(uint32_t address, uint32_t size);
+
+/***************************************************************************************************
+** PRIVATE FUNCTIONS
+***************************************************************************************************/
+
+/*!
+** @brief Gets the FLASH sector of an address.
+** @return Returns the FLASH sector if valid and 10 if a non-valid address is supplied.
+*/
+static uint32_t getFlashSector(uint32_t address)
+{
+    uint32_t address_sectors_start[VALID_SECTORS] = {START_SECTOR_1, START_SECTOR_2, START_SECTOR_3,
+                                                     START_SECTOR_4, START_SECTOR_5};
+
+    uint32_t address_sectors_end[VALID_SECTORS] = {END_SECTOR_1, END_SECTOR_2, END_SECTOR_3,
+                                                   END_SECTOR_4, END_SECTOR_5};                                         
+    
+    for (int i = 0; i < VALID_SECTORS; i++)
+    {
+        if (address >= address_sectors_start[i] &&
+            address <= address_sectors_end[i])
+        {
+            return i+1;
+        }
+    }
+    // This should never occur - return non-existing sector.
+    return 10;
+}
+
+/*!
+** @brief Checks that the data to be stored is contained within a single FLASH sector.
+** @return Returns 0 if the data is contained within a single FLASH sector, otherwise 1.
+*/
+static int isWriteWithinSector(uint32_t address, uint32_t size)
+{
+    uint32_t sector_start = getFlashSector(address);
+    uint32_t sector_end = getFlashSector(address + size);
+
+    // Ensure that the start and end addresses belong to the same FLASH sector.
+    if (sector_start != sector_end)
+    {
+        return 1;
+    }
+    return 0;
+}
+
+/*!
+** @brief Checks whether the user is trying to write to a memory range containing program memory.
+** @return Returns 1 if the memory range contains program memory, otherwise 0.
+*/
+static int isProgramMemory(uint32_t address, uint32_t size)
+{
+    if (address >= PROGRAM_START_ADDR && address < PROGRAM_END_ADDR)
```

Comment:
<=? (what timothe said)

Reply:
See comment above. I will add a comment explaining this.